### PR TITLE
Verify bundle resources do not conflict

### PR DESCRIPTION
**Pull request by rmaz**
Make it a build error when an apple bundle has multiple resources with the same path that will end up overwriting each other.

### CODE REVIEW

**Comment by robbertvanginkel** [comment]
@milend can you take a look at this?

File: src/com/facebook/buck/apple/AppleBundle.java [hunk 1]
```diff
@@ -426,6 +427,7 @@ public boolean isLegacyWatchApp() {
             resources.getResourceDirs(),
             resources.getDirsContainingResourceDirs(),
             resources.getResourceFiles()))) {
+      verifyResourceConflicts(resources, context.getSourcePathResolver());
```
Comment (milend) [inline src/com/facebook/buck/apple/AppleBundle.java]: We should introduce this behaviour behind a config option as this has the potential to break existing apps which already include duplicate resources (which means we cannot update Buck without fixing any offending apps which is impractical).

The config for this should be off by default.

File: src/com/facebook/buck/apple/AppleBundle.java [hunk 2]
```diff
@@ -670,6 +672,24 @@ public boolean isLegacyWatchApp() {
     return stepsBuilder.build();
   }
 
+  private void verifyResourceConflicts(AppleBundleResources resources,
+                                       SourcePathResolver resolver) {
+    // Ensure there are no resources that will overwrite each other
+    Set<Path> filenames = new HashSet<>();
+    for (SourcePath path : Iterables.concat(
+        resources.getResourceDirs(),
```
Comment (milend) [inline src/com/facebook/buck/apple/AppleBundle.java]: It seems that this is duplicating the resource paths form lines 427-429:

    resources.getResourceDirs(),
    resources.getDirsContainingResourceDirs(),
    resources.getResourceFiles()

We should extract those and pass them into the method instead.

File: src/com/facebook/buck/apple/AppleBundle.java [hunk 3]
```diff
@@ -670,6 +672,24 @@ public boolean isLegacyWatchApp() {
...
+        resources.getDirsContainingResourceDirs(),
+        resources.getResourceFiles())) {
+      Path filename = resolver.getRelativePath(path).getFileName();
+      if (filenames.contains(filename)) {
+        throw new HumanReadableException(
+            "Bundle contains multiple resources with filename %s", filename);
```
Comment (milend) [inline src/com/facebook/buck/apple/AppleBundle.java]: Nit: `filename` can be inaccurate as it can be directories, too. Can we improve the naming?

File: src/com/facebook/buck/apple/AppleBundle.java [hunk 4]
```diff
@@ -670,6 +672,24 @@ public boolean isLegacyWatchApp() {
     return stepsBuilder.build();
   }
 
+  private void verifyResourceConflicts(AppleBundleResources resources,
+                                       SourcePathResolver resolver) {
+    // Ensure there are no resources that will overwrite each other
+    Set<Path> filenames = new HashSet<>();
+    for (SourcePath path : Iterables.concat(
+        resources.getResourceDirs(),
+        resources.getDirsContainingResourceDirs(),
```
Comment (milend) [inline src/com/facebook/buck/apple/AppleBundle.java]: This check won't work correctly because it's the contents *inside* that get copied to the root of the resources dir, not the dirs themselves.

The deeper issue here is that those dirs will usually not have the contents at this point but only be generated later (e.g., a `genrule`). So you won't be able to perform the check here.

Given the above, you'd have to move the check to be part of the `CopyStep` itself.

File: test/com/facebook/buck/apple/AppleBundleIntegrationTest.java [hunk 5]
```diff
@@ -622,6 +622,16 @@ public void appBundleWithResources() throws Exception {
             "%s"));
   }
 
+  @Test
+  public void appBundleWithConflictingResources() throws Exception {
```
Comment (milend) [inline test/com/facebook/buck/apple/AppleBundleIntegrationTest.java]: Can you add another test where we have duplicate nested filenames, so something like:

    FolderA/Image.png
    FolderB/Image.png

Comment (rmaz) [reply test/com/facebook/buck/apple/AppleBundleIntegrationTest.java]: How can I add a test to cover `DirsContainingResourceDirs` too?

File: src/com/facebook/buck/apple/AppleBundle.java (hunk 4 above)
Comment (rmaz) [reply src/com/facebook/buck/apple/AppleBundle.java]: I was a bit confused about `DirsContainingResourceDirs`, is this only the output of genrules or also other stuff? How can I set up a BUCK config to check this?

Comment (rmaz) [reply src/com/facebook/buck/apple/AppleBundle.java]: If we move this to `CopyStep`, do you want this to be enabled globally or just for apple bundle resources?

Comment (milend) [reply src/com/facebook/buck/apple/AppleBundle.java]: @rmaz I had to look into it and it seems it's only used for `js_bundle` (see `JsBundleDescription::addAppleBundleResources()`). I couldn't find any integration tests for `js_bundle`, so I think we can ignore `DirsContainingResourceDirs` for now (but please put a `TODO`).

File: test/com/facebook/buck/apple/AppleBundleIntegrationTest.java (hunk 5 above)
Comment (milend) [reply test/com/facebook/buck/apple/AppleBundleIntegrationTest.java]: @rmaz As per my above comment, I don't think we can easily test this as `js_bundle` doesn't seem to have integration tests.

File: src/com/facebook/buck/apple/AppleBundle.java (hunk 1 above)
Comment (rmaz) [reply src/com/facebook/buck/apple/AppleBundle.java]: done

**Comment by rmaz** [comment]
Updated per comments, if you have some preferred naming let me know

File: test/com/facebook/buck/apple/AppleBundleIntegrationTest.java [hunk 6]
```diff
@@ -632,6 +632,16 @@ public void appBundleWithConflictingResources() throws Exception {
...
   }
 
+  @Test
+  public void appBundleWithConflictingNestedFolderResources() throws Exception {
+    ProjectWorkspace workspace =
+        TestDataHelper.createProjectWorkspaceForScenario(this, "app_bundle_with_conflicting_nested_resources", tmp);
+    workspace.setUp();
+
+    BuildTarget target = workspace.newBuildTarget("//:DemoApp#iphonesimulator-x86_64,no-debug");
+    workspace.runBuckCommand("build", target.getFullyQualifiedName()).assertFailure();
```
Comment (milend) [inline test/com/facebook/buck/apple/AppleBundleIntegrationTest.java]: @rmaz We should add another test that actually tests for success but has files with the same name inside the folders. We can just copy `app_bundle_with_conflicting_nested_resources` and change:

    files = glob(["**/*.png"]),

to:

    dirs = ["ImagesA", "ImagesB"],

Alternatively, we can just skip the images as the localizable strings all have the same file names inside.

**Comment by milend** [comment]
@rmaz LGTM, I'd just want to add one more test that asserts success (in addition to the other two tests that assert failure).

**Comment by rmaz** [comment]
done, is this what you were after?

**Comment by milend** [comment]
@rmaz Yup!
